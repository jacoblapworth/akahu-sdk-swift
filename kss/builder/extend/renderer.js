const { regexDictionary, types } = require('../parser/constants');
const tinycolor = require('tinycolor2');

module.exports = function(handlebars) {
	/**
	 * Extracts the variable properties from the line
	 * returns:
	 *  token name, token value, and optional [ token example ]
	 */
	handlebars.registerHelper('renderTokens', renderTokens);

	/**
	 * Compares the name to see if the container needs to be rendered as inverted.
	 * Similar to the isEqual
	 *
	 * e.g.
	 * {{#isInverted name}}
	 * 		ADD THIS
	 * {{else}} [optional]
	 * 		WILL EXECUTE THIS
	 * {{/isInverted}}
	 */
	handlebars.registerHelper('isInverted', isInverted);
};

const testString = 'To understand what recursion is, you must first understand recursion.';


function renderTokens(variables, block) {
	var newBlock = '';
	variables.split('\n').forEach(line => {
		var test;
		const matches = line.match(/[^:]+/g).map(match => match.trim());

		if (matches[1]) {
			this.token = {};
			this.token.name = matches[0];
			this.token.value = matches[1];
			this.token.title = matches[3];
			this.token.description = matches[4];
			if (matches[2] != undefined) {
				this.token.example = getExample(matches[2], this.token);
			}
			newBlock += block.fn(this);
		}
	});
	return newBlock;
}

function getExample(type, token) {
	switch(type) {
		case types.color:
			const codeTextClass = tinycolor(token.value).isDark() ?
				'ds-colour-code-light'
				: null;
			return `
					<div class="ds-colour-example" style="background-color:${token.value}">
						<code class="${codeTextClass}">${token.value}</code>
						<code class="${codeTextClass}">${token.name}</code>
					</div>`;
		case types.border:
			return `
				<div class="ds-colour-example" style="border:${token.value}">
					<code>${token.name}</code>
				</div>`;
		case types.shadow:
			return `
					<div class="ds-colour-example" style="box-shadow:${token.value}">
						<code>${token.name}</code>
					</div>`;
		case types.fontSize:
			return  `<span class="xui-text-truncated" style="font-size:${token.value};">Aa</span>`;
		case types.fontWeight:
			return `<span style="font-weight:${token.value};">Aa</span>`;
		case types.lineHeight:
			return `<span style="line-height:${token.value}">${testString}</span>`;
		case types.spacing:
			return `<div class="ds-spacing-example" style="height:${token.value};width:${token.value};"></div>`;
		default:
			return '';
	}
}

function isInverted(name, options) {
	return regexDictionary.inverted.test(name) ? options.fn(this) : options.inverse(this);
}
