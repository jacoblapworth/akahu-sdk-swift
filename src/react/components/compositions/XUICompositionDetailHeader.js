import React, { PureComponent } from 'react';
import PropTypes from 'prop-types';
import cn from 'classnames';

import XUIGridAreaDetail from './XUIGridAreaDetail';
import XUIGridAreaHeader from './XUIGridAreaHeader';

import baseCompositionClass, { buildGlobalCompositionClasses } from './helpers';

export default class XUICompositionDetailHeader extends PureComponent {
  render() {
    const { className, header, detail, ...spreadProps } = this.props;

    const compositionClasses = cn(
      buildGlobalCompositionClasses(spreadProps),
      `${baseCompositionClass}-detailheader`,
      className,
    );

    return (
      <div className={compositionClasses}>
        <XUIGridAreaHeader>{header}</XUIGridAreaHeader>
        <XUIGridAreaDetail>{detail}</XUIGridAreaDetail>
      </div>
    );
  }
}

XUICompositionDetailHeader.propTypes = {
  className: PropTypes.string,

  /**
   * Header content or component
   */
  header: PropTypes.element.isRequired,
  /**
   * Main content
   */
  detail: PropTypes.element.isRequired,
  /**
   * Determines whether the main content takes full width of page. Defaults to false.
   */
  isInfinite: PropTypes.bool,
  /**
   * Whether to apply a pre-set grid-gap between all grid areas. Defaults to true.
   */
  hasGridGap: PropTypes.bool,
  /**
   * Whether to apply pre-set spacing to the outside of the composition grid. Defaults to true.
   */
  hasAutoSpaceAround: PropTypes.bool,
};

XUICompositionDetailHeader.defaultProps = {
  hasGridGap: true,
  hasAutoSpaceAround: true,
};
