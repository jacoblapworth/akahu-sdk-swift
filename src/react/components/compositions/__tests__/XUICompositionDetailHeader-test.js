import React from 'react';
import Enzyme from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import renderer from 'react-test-renderer';

import XUICompositionDetailHeader from '../XUICompositionDetailHeader';

Enzyme.configure({ adapter: new Adapter() });

describe('<XUICompositionDetailHeader>', () => {

	it('renders basic example', () => {
		const wrapper = renderer.create(
			<XUICompositionDetailHeader
				header={<div></div>}
				main={<div></div>}
				/>
		);
		expect(wrapper).toMatchSnapshot();
	});

	it('should do nothing with children', () => {
		const wrapper = renderer.create(
			<XUICompositionDetailHeader
				header={<div></div>}
				main={<div></div>}
				>
				Hello
			</XUICompositionDetailHeader>
		);
		expect(wrapper).toMatchSnapshot();
	});
	it('should include custom class and omit width-limiting class, if specified', () => {
		const wrapper = renderer.create(
			<XUICompositionDetailHeader
				header={<div></div>}
				main={<div></div>}
				className="single-with-head"
				isInfinite={true}
				/>
		);
		expect(wrapper).toMatchSnapshot();
	});
});
