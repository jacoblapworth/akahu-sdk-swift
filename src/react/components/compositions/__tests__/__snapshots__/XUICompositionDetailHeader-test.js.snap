// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<XUICompositionDetailHeader> renders basic example 1`] = `
<div
  className="xui-composition xui-composition-is-finite xui-composition-has-grid-gap xui-composition-has-auto-space-around xui-composition-detailheader"
>
  <div
    className="xui-gridarea-header"
  >
    <div />
  </div>
  <div
    className="xui-gridarea-detail"
  >
    <div />
  </div>
</div>
`;

exports[`<XUICompositionDetailHeader> should do nothing with children 1`] = `
<div
  className="xui-composition xui-composition-is-finite xui-composition-has-grid-gap xui-composition-has-auto-space-around xui-composition-detailheader"
>
  <div
    className="xui-gridarea-header"
  >
    <div />
  </div>
  <div
    className="xui-gridarea-detail"
  >
    <div />
  </div>
</div>
`;

exports[`<XUICompositionDetailHeader> should include custom class and toggle off default classes, if specified 1`] = `
<div
  className="xui-composition xui-composition-detailheader single-with-head"
>
  <div
    className="xui-gridarea-header"
  >
    <div />
  </div>
  <div
    className="xui-gridarea-detail"
  >
    <div />
  </div>
</div>
`;
